### ISSUE TITLE
SOFT ASSERT FAILED when using rlm_perl

### ISSUE THREAD

**Issue by citywok** [FreeRADIUS/freeradius-server#1807]
# Issue type
- [ ] Defect - Crash or memory corruption.
- [ ] Defect - Non compliance with a standards document, or incorrect API usage.
- [X] Defect - Unexpected behaviour (obvious or verified by project member).
- [ ] Feature request.

Potentially related to https://github.com/FreeRADIUS/freeradius-server/issues/1338

# Defect/Feature description
## How to reproduce issue
Using FreeRADIUS 3.0.12 with rlm_sql, rlm_perl, and eap_peap.  We store the password encrypted in MySQL, then hand the encrypted-password to perl which decrypts it and then returns it as cleartext-password so that freeradius can compare to what the client provided.  This perl process has been working in 3.0.11 for 6 months but now that we've upgraded to 3.0.12 we see several SOFT ASSERT FAILED messages in the logs for each authentication.

We only see the soft assert failed messages when we are using rlm_perl, for authentications that do not go through perl we do not see the soft assert failed message.

Note that the authentications succeed even though radius is saying soft assert failed in the logs.
 
server inner-tunnel {
authorize {
    mschap
    suffix
    eap
    if ("%{Called-Station-Id}" =~ /wireless_ssid/){
        update {
            SSID = 'wireless_ssid'
        }
        redundant {
            sql_primary
            sql_secondary
        }
        perl
    }
.....
}

## Output of ``[radiusd|freeradius] -X`` showing issue occurring

```
(158) # Executing group from file /environment_root/var/etc/raddb/sites-enabled/default
(158)   Auth-Type eap {
(158) eap: Expiring EAP session with state 0xbf4a6930ba4c70e2
(158) eap: Finished EAP session with state 0xbf4a6930ba4c70e2
(158) eap: Previous EAP request found for state 0xbf4a6930ba4c70e2, released from the list
(158) eap: Peer sent packet with method EAP PEAP (25)
(158) eap: Calling submodule eap_peap to process data
(158) eap_peap: Continuing EAP-TLS
(158) eap_peap: [eaptls verify] = ok
(158) eap_peap: Done initial handshake
(158) eap_peap: [eaptls process] = ok
(158) eap_peap: Session established.  Decoding tunneled attributes
(158) eap_peap: PEAP state WAITING FOR INNER IDENTITY
(158) eap_peap: Identity - peap_username@myrealm
(158) eap_peap: Got inner identity 'peap_username@myrealm'
(158) eap_peap: Setting default EAP type for tunneled EAP session
(158) eap_peap: Got tunneled request
(158) eap_peap:   EAP-Message = 0x02060016017363616e6e65722d31407363616e6e6572
(158) eap_peap: Setting User-Name to peap_username@myrealm
(158) eap_peap: Sending tunneled request to inner-tunnel
(158) eap_peap:   EAP-Message = 0x02060016017363616e6e65722d31407363616e6e6572
(158) eap_peap:   FreeRADIUS-Proxied-To = 127.0.0.1
(158) eap_peap:   User-Name = "peap_username@myrealm"
(158) eap_peap:   NAS-IP-Address = 127.0.0.1
(158) eap_peap:   Calling-Station-Id = "02-00-00-00-00-01"
(158) eap_peap:   Framed-MTU = 1400
(158) eap_peap:   NAS-Port-Type = Wireless-802.11
(158) eap_peap:   Connect-Info = "CONNECT 11Mbps 802.11b"
(158) eap_peap:   Called-Station-Id = "wireless_ssid"
(158) eap_peap:   NAS-Identifier = "fake-identifier"
(158) eap_peap:   Event-Timestamp = "Oct 25 2016 22:49:46 UTC"
(158) Virtual server inner-tunnel received request
(158)   EAP-Message = 0x02060016017363616e6e65722d31407363616e6e6572
(158)   FreeRADIUS-Proxied-To = 127.0.0.1
(158)   User-Name = "peap_username@myrealm"
(158)   NAS-IP-Address = 127.0.0.1
(158)   Calling-Station-Id = "02-00-00-00-00-01"
(158)   Framed-MTU = 1400
(158)   NAS-Port-Type = Wireless-802.11
(158)   Connect-Info = "CONNECT 11Mbps 802.11b"
(158)   Called-Station-Id = "wireless_ssid"
(158)   NAS-Identifier = "fake-identifier"
(158)   Event-Timestamp = "Oct 25 2016 22:49:46 UTC"
(158) WARNING: Outer and inner identities are the same.  User privacy is compromised.
(158) server inner-tunnel {
(158)   # Executing section authorize from file /environment_root/var/etc/raddb/sites-enabled/inner-tunnel
(158)     authorize {
(158)       [mschap] = noop
(158) suffix: Checking for suffix after "@"
(158) suffix: Looking up realm "scanner" for User-Name = "peap_username@myrealm"
(158) suffix: Found realm "DEFAULT"
(158) suffix: Adding Stripped-User-Name = "peap_username"
(158) suffix: Adding Realm = "DEFAULT"
(158) suffix: Authentication realm is LOCAL
(158)       [suffix] = ok
(158) eap: Peer sent EAP Response (code 2) ID 6 length 22
(158) eap: EAP-Identity reply, returning 'ok' so we can short-circuit the rest of authorize
(158)       [eap] = ok
(158)       if ("%{Called-Station-Id}" =~ /wireless_ssid/){
(158)       EXPAND %{Called-Station-Id}
(158)          --> wireless_ssid
(158)       if ("%{Called-Station-Id}" =~ /wireless_ssid/) -> TRUE
(158)       if ("%{Called-Station-Id}" =~ /wireless_ssid/) {
(158)         update {
(158)           SSID = 'wireless_ssid'
(158)         } # update = noop
(158)         redundant {
(158) sql_primary: EXPAND %{SSID}.%{User-Name}
(158) sql_primary:    --> wireless_ssid.peap_username@myrealm
(158) sql_primary: SQL-User-Name set to 'wireless_ssid.peap_username@myrealm'
rlm_sql (sql_primary): Reserved connection (5)
(158) sql_primary: EXPAND SELECT id, username, attribute, value, op FROM radcheck WHERE username = '%{SQL-User-Name}' ORDER BY id
(158) sql_primary:    --> SELECT id, username, attribute, value, op FROM radcheck WHERE username = 'wireless_ssid.peap_username@myrealm' ORDER BY id
(158) sql_primary: Executing select query: SELECT id, username, attribute, value, op FROM radcheck WHERE username = 'wireless_ssid.peap_username@myrealm' ORDER BY id
(158) sql_primary: User found in radcheck table
(158) sql_primary: Conditional check items matched, merging assignment check items
(158) sql_primary:   Encrypted-Password := "*masked*"
(158) sql_primary: EXPAND SELECT id, username, attribute, value, op FROM radreply WHERE username = '%{SQL-User-Name}' ORDER BY id
(158) sql_primary:    --> SELECT id, username, attribute, value, op FROM radreply WHERE username = 'wireless_ssid.peap_username@myrealm' ORDER BY id
(158) sql_primary: Executing select query: SELECT id, username, attribute, value, op FROM radreply WHERE username = 'wireless_ssidpeap_username@myrealm' ORDER BY id
(158) sql_primary: User found in radreply table, merging reply items
(158) sql_primary:   Airespace-Interface-Name := "airespace-1"
rlm_sql (sql_primary): Released connection (5)
rlm_sql (sql_primary): Need 2 more connections to reach 3 spares
rlm_sql (sql_primary): Opening additional connection (7), 1 of 126 pending slots used
rlm_sql_mysql: Starting connect to MySQL server
rlm_sql_mysql: Connected to database 'freeradius' on hostname.tld via TCP/IP, server version 5.6.27, protocol version 10
(158)           [sql_primary] = ok
(158)         } # redundant = ok
(158) perl:   $RAD_REQUEST{'User-Name'} = &request:User-Name -> 'peap_username@myrealm'
(158) perl:   $RAD_REQUEST{'NAS-IP-Address'} = &request:NAS-IP-Address -> '127.0.0.1'
(158) perl:   $RAD_REQUEST{'Framed-MTU'} = &request:Framed-MTU -> '1400'
(158) perl:   $RAD_REQUEST{'Called-Station-Id'} = &request:Called-Station-Id -> 'wireless_ssid'
(158) perl:   $RAD_REQUEST{'Calling-Station-Id'} = &request:Calling-Station-Id -> '02-00-00-00-00-01'
(158) perl:   $RAD_REQUEST{'NAS-Identifier'} = &request:NAS-Identifier -> 'fake-identifier'
(158) perl:   $RAD_REQUEST{'NAS-Port-Type'} = &request:NAS-Port-Type -> 'Wireless-802.11'
(158) perl:   $RAD_REQUEST{'Event-Timestamp'} = &request:Event-Timestamp -> 'Oct 25 2016 22:49:46 UTC'
(158) perl:   $RAD_REQUEST{'Connect-Info'} = &request:Connect-Info -> 'CONNECT 11Mbps 802.11b'
(158) perl:   $RAD_REQUEST{'EAP-Message'} = &request:EAP-Message -> '0x02060016017363616e6e65722d31407363616e6e6572'
(158) perl:   $RAD_REQUEST{'FreeRADIUS-Proxied-To'} = &request:FreeRADIUS-Proxied-To -> '127.0.0.1'
(158) perl:   $RAD_REQUEST{'EAP-Type'} = &request:EAP-Type -> 'Identity'
(158) perl:   $RAD_REQUEST{'Stripped-User-Name'} = &request:Stripped-User-Name -> 'peap_username'
(158) perl:   $RAD_REQUEST{'Realm'} = &request:Realm -> 'DEFAULT'
(158) perl:   $RAD_REQUEST{'SSID'} = &request:SSID -> 'wireless_ssid'
(158) perl:   $RAD_REPLY{'Airespace-Interface-Name'} = &reply:Airespace-Interface-Name -> 'airespace-1'
(158) perl:   $RAD_CHECK{'Auth-Type'} = &control:Auth-Type -> 'eap'
(158) perl:   $RAD_CHECK{'EAP-Type'} = &control:EAP-Type -> 'MSCHAPv2'
(158) perl:   $RAD_CHECK{'Encrypted-Password'} = &control:Encrypted-Password -> '*masked*'
(158) perl:   $RAD_CONFIG{'Auth-Type'} = &control:Auth-Type -> 'eap'
(158) perl:   $RAD_CONFIG{'EAP-Type'} = &control:EAP-Type -> 'MSCHAPv2'
(158) perl:   $RAD_CONFIG{'Encrypted-Password'} = &control:Encrypted-Password -> '*masked*'
SOFT ASSERT FAILED /original_build_path/build/private/src/src/modules/rlm_perl/rlm_perl.c[703]: *vps
(158) perl: &request:NAS-Port-Type = $RAD_REQUEST{'NAS-Port-Type'} -> 'Wireless-802.11'
(158) perl: &request:SSID = $RAD_REQUEST{'SSID'} -> 'wireless_ssid'
(158) perl: &request:Calling-Station-Id = $RAD_REQUEST{'Calling-Station-Id'} -> '02-00-00-00-00-01'
(158) perl: &request:Called-Station-Id = $RAD_REQUEST{'Called-Station-Id'} -> 'wireless_ssid'
(158) perl: &request:FreeRADIUS-Proxied-To = $RAD_REQUEST{'FreeRADIUS-Proxied-To'} -> '127.0.0.1'
(158) perl: &request:User-Name = $RAD_REQUEST{'User-Name'} -> 'peap_username@myrealm'
(158) perl: &request:NAS-Identifier = $RAD_REQUEST{'NAS-Identifier'} -> 'fake-identifier'
(158) perl: &request:Event-Timestamp = $RAD_REQUEST{'Event-Timestamp'} -> 'Oct 25 2016 22:49:46 UTC'
(158) perl: &request:EAP-Message = $RAD_REQUEST{'EAP-Message'} -> '0x02060016017363616e6e65722d31407363616e6e6572'
(158) perl: &request:Connect-Info = $RAD_REQUEST{'Connect-Info'} -> 'CONNECT 11Mbps 802.11b'
(158) perl: &request:Realm = $RAD_REQUEST{'Realm'} -> 'DEFAULT'
(158) perl: &request:Stripped-User-Name = $RAD_REQUEST{'Stripped-User-Name'} -> 'peap_username'
(158) perl: &request:EAP-Type = $RAD_REQUEST{'EAP-Type'} -> 'Identity'
(158) perl: &request:NAS-IP-Address = $RAD_REQUEST{'NAS-IP-Address'} -> '127.0.0.1'
(158) perl: &request:Framed-MTU = $RAD_REQUEST{'Framed-MTU'} -> '1400'
SOFT ASSERT FAILED /original_build_path/build/private/src/src/modules/rlm_perl/rlm_perl.c[703]: *vps
(158) perl: &reply:Airespace-Interface-Name = $RAD_REPLY{'Airespace-Interface-Name'} -> 'airespace-1'
SOFT ASSERT FAILED /original_build_path/build/private/src/src/modules/rlm_perl/rlm_perl.c[703]: *vps
(158) perl: &control:Cleartext-Password = $RAD_CHECK{'Cleartext-Password'} -> '*masked*'
(158) perl: &control:Auth-Type = $RAD_CHECK{'Auth-Type'} -> 'eap'
(158) perl: &control:EAP-Type = $RAD_CHECK{'EAP-Type'} -> 'MSCHAPv2'
(158)         [perl] = updated
(158)       } # if ("%{Called-Station-Id}" =~ /wireless_ssid/) = updated
(158)       ... skipping elsif: Preceding "if" was taken
(158)       ... skipping else: Preceding "if" was taken
(158) pap: WARNING: Auth-Type already set.  Not setting to PAP
(158)       [pap] = noop
(158)     } # authorize = updated
(158)   Found Auth-Type = eap
(158)   # Executing group from file /environment_root/var/etc/raddb/sites-enabled/inner-tunnel
(158)     Auth-Type eap {
(158) eap: Peer sent packet with method EAP Identity (1)
(158) eap: Calling submodule eap_mschapv2 to process data
(158) eap_mschapv2: Issuing Challenge
(158) eap: Sending EAP Request (code 1) ID 7 length 43
(158) eap: EAP session adding &reply:State = 0xe9e93749e9ee2dbc
(158)       [eap] = handled
(158)     } # Auth-Type eap = handled
```

**Comment by arr2036** [comment]
Already fixed, use v3.0.x

**Comment by citywok** [comment]
Sweet, I'll give it a shot.  Do you know when the release of 3.0.13 will be?

Thanks!
